File: research/tise_research/features/attention.py
"""Is attention predictable? Dwell labels and features, one per visit.

> Given a page you have just opened, will you stay longer than you usually stay on pages of
> this category?

**This is the `full` compat class, and that is the point of measuring it now.** Since T1 the
duration trap has been recorded as permanent: the history *file* has `visit_duration`, the
`chrome.history` API does not, so every dwell analysis was research-only and unshippable.
`chrome.tabs` would let the extension measure dwell live — so the question stops being
academic, and the file lets us answer it **before** asking for that permission rather than
after.

**Why a visit is the right unit.** `return_24h` had one label per (category, session);
`block_volume` had one per (category, day) and produced 403 across three corpora. A visit
gives one label per *event* — thousands. Label scarcity has been the binding constraint on
every target this project has tried, and this is the finest unit the data contains.

**Why the label is a median split against the category's own recent dwell.** It makes the
base rate ~50% for any person and any category by construction — the property D88 wanted and
D90 measured as false for daily blocks, where the unit was coarse enough for volume trends to
dominate. At visit level the trailing window spans hours rather than weeks, so there is far
less room for drift.

**No URL, no path.** `Visit` drops the URL at the reader boundary and nothing here reaches
past it. Path-derived features are a separate decision that has not been made.

Leakage is structural, as in `block_labels.py`: a visit's label and features are emitted
before that visit joins any state.

**Two feature sets, one label.** `as_1` is what D93 measured. `as_2` adds domain and
sequence features and is what T-A tests, and the label is deliberately **identical** so any
change in score is attributable to the features and the cluster unit rather than to a
redefinition. Both are computed in one pass and the requested set is selected, following
`compute_features`: recomputing the traversals per set would be the same work twice with
two chances to disagree. Regenerating the `as_1` report must move no figure, which is a
check on that claim rather than a hope.
"""

from __future__ import annotations

import math
from collections import deque
from collections.abc import Sequence
from dataclasses import dataclass
from datetime import date, timedelta
from statistics import median

from tise_research.features.events import Event
from tise_research.features.labels import Label
from tise_research.features.sessions import sessionise
from tise_research.features.vector import FeatureRow, feature_names, saturate

__all__ = [
    "ATTENTION_FEATURE_SET",
    "DAILY_DOMAIN_MIN_DAYS",
    "DAILY_DOMAIN_WINDOW_DAYS",
    "DEFAULT_MIN_PRIOR_VISITS",
    "DEFAULT_TRAILING_VISITS",
    "DOMAIN_SHARE_WINDOW",
    "DOMAIN_VISIT_SCALE",
    "ENGAGED_FEATURE_SET",
    "VISIT_ENGAGED_TARGET",
    "AttentionExample",
    "attention_examples",
]

#: One name for one thing. SPEC.md's `Prediction.target` union has said `visit_engaged`
#: since D94 while this constant said `attention_dwell`, which is two names for a single
#: target and exactly how a divergence starts — D87 found the same shape three times. The
#: label is unchanged; only what it is called is. No published figure moves, because no
#: report prints the target string.
VISIT_ENGAGED_TARGET = "visit_engaged"

#: What D93 measured. Kept so that report stays reproducible.
ATTENTION_FEATURE_SET = "as_1"

#: What T-A measures. Same label, more features — see the module docstring.
ENGAGED_FEATURE_SET = "as_2"

#: Visits of the same category the median is taken over.
DEFAULT_TRAILING_VISITS = 20

#: Visits of a category before it can be predicted at all.
DEFAULT_MIN_PRIOR_VISITS = 10

#: Dwell at which `dwellLevel` reaches 0.5, in seconds. Thirty seconds is an ordinary page
#: read; declared, not fitted, per D81.
DWELL_SCALE_SECONDS = 30.0

#: Session position saturation, in visits. Declared.
POSITION_SCALE = 10.0

#: Prior visits at which `domainVisits` reaches 0.5. Twenty, matching the trailing window
#: the median is taken over, so there is one notion of "enough history" and not two.
#: Declared, not fitted.
DOMAIN_VISIT_SCALE = 20.0

#: Visits `domainShare` is taken over. A hundred is a few days of this person's browsing —
#: long enough that a share is not one visit's worth of noise, short enough that a domain
#: abandoned last month has left it. Declared.
DOMAIN_SHARE_WINDOW = 100

#: `isDailyDomain` looks back this many calendar days and fires at this many of them.
#: Four of seven is a majority of the week, which is what "daily" is being used to mean.
#: Declared.
DAILY_DOMAIN_WINDOW_DAYS = 7
DAILY_DOMAIN_MIN_DAYS = 4


@dataclass(frozen=True, slots=True)
class AttentionExample:
    label: Label
    row: FeatureRow
    #: The domain this visit was to. Carried so a per-domain **baseline** can be fitted —
    #: D24 makes baselines mandatory, and once four of `as_2`'s six features read the
    #: domain, "a per-domain rate table would do the same job" is the first thing a reader
    #: should be able to check. It is not on the `Label`, because a label is a question and
    #: the domain is an input to answering it.
    #:
    #: **Never published.** It stays inside the research tier: reports may print counts of
    #: domains, never a domain. Invariant 2 is enforced upstream — `Visit` drops the URL at
    #: the reader boundary — and this is the reduced host, which `Event` has carried since
    #: T1 and no report has ever printed.
    domain: str = ""


def _transition_flags(transition: str) -> dict[str, float]:
    """How the person arrived. The one genuinely *conditional* fact each visit carries.

    `typed` and `auto_bookmark` are deliberate acts; `link` is incidental. Nothing in this
    project has ever used the distinction, though it has been collected since T1.
    """
    core = transition.lower()
    return {
        "arrivedTyped": 1.0 if core in {"typed", "generated", "keyword"} else 0.0,
        "arrivedBookmark": 1.0 if "bookmark" in core else 0.0,
        "arrivedLink": 1.0 if core == "link" else 0.0,
    }


def attention_examples(
    events: Sequence[Event],
    *,
    timeout_seconds: float,
    trailing: int = DEFAULT_TRAILING_VISITS,
    min_prior: int = DEFAULT_MIN_PRIOR_VISITS,
    feature_set: str = ATTENTION_FEATURE_SET,
) -> list[AttentionExample]:
    """One example per visit that has enough history of its own category.

    Visits with no recorded dwell are skipped rather than imputed: a missing duration is not
    a short one, and inventing a value here would be the defect D51 forbids.

    `feature_set` selects `as_1` or `as_2`. **It changes only the features.** The label —
    outcome, subject, `window_end`, and which visits get one at all — is identical either
    way, which is what makes D93's numbers and T-A's comparable (D94).
    """
    names = feature_names(feature_set)
    sessions = sessionise(events, timeout_seconds=timeout_seconds)

    history: dict[str, list[float]] = {}
    domain_dwell: dict[str, list[float]] = {}
    domain_visits: dict[str, int] = {}
    domain_days: dict[str, set[date]] = {}
    #: Bounded, so `domainShare` describes recent browsing rather than all of it.
    recent_domains: deque[str] = deque(maxlen=DOMAIN_SHARE_WINDOW)

    examples: list[AttentionExample] = []
    seen: set[str] = set()
    previous_category: str | None = None
    previous_domain: str | None = None
    previous_dwell: float | None = None

    for session in sessions:
        for position, event in enumerate(session.events):
            dwell = event.dwell_seconds
            prior = history.get(event.category, [])
            window = prior[-trailing:]

            if dwell is not None and len(prior) >= min_prior:
                threshold = median(window)
                hour_angle = 2.0 * math.pi * event.occurred_at.hour / 24.0

                # --- domain state, all of it strictly before this visit ---
                domain_window = domain_dwell.get(event.domain, [])[-trailing:]
                first_day = event.occurred_at.date() - timedelta(
                    days=DAILY_DOMAIN_WINDOW_DAYS - 1
                )
                days_recently = sum(
                    1 for day in domain_days.get(event.domain, ()) if day >= first_day
                )
                # Guaranteed non-empty: `min_prior` prior visits of this category are
                # themselves prior visits. Asserted rather than given a zero fallback,
                # because a share over nothing is undefined and not zero.
                assert recent_domains, "a labelled visit always has prior visits"

                values: dict[str, float | None] = {
                    "dwellLevel": saturate(float(threshold), DWELL_SCALE_SECONDS),
                    "lastDwellRatio": window[-1] / (threshold + 1.0),
                    "meanRatio": (sum(window) / len(window)) / (threshold + 1.0),
                    "sessionPosition": saturate(float(position), POSITION_SCALE),
                    "isSessionStart": 1.0 if position == 0 else 0.0,
                    "sameAsPrevious": (
                        1.0 if previous_category == event.category else 0.0
                    ),
                    "hourSin": math.sin(hour_angle),
                    "hourCos": math.cos(hour_angle),
                    "isWeekend": 1.0 if event.occurred_at.weekday() >= 5 else 0.0,
                    **_transition_flags(event.transition),
                    # --- as_2 ---
                    "domainVisits": saturate(
                        float(domain_visits.get(event.domain, 0)), DOMAIN_VISIT_SCALE
                    ),
                    "domainShare": (
                        sum(1 for name in recent_domains if name == event.domain)
                        / len(recent_domains)
                    ),
                    "isDailyDomain": (
                        1.0 if days_recently >= DAILY_DOMAIN_MIN_DAYS else 0.0
                    ),
                    "domainDwellLevel": (
                        saturate(float(median(domain_window)), DWELL_SCALE_SECONDS)
                        if domain_window
                        else None
                    ),
                    "prevSameDomain": 1.0 if previous_domain == event.domain else 0.0,
                    "prevDwellRatio": (
                        None
                        if previous_dwell is None
                        else previous_dwell / (threshold + 1.0)
                    ),
                }

                # The key consumers index rows by. It is the *event* id, not the
                # (category, instant) pair those consumers used until D99: that pair is
                # unique on Akash's browsing and not unique in general — the GESIS panel
                # records to the second, and one person can visit two pages of a category
                # within one. The guard stays, now on the thing that actually has to be
                # unique, because a duplicate here silently scores one row against two
                # labels.
                if event.event_id in seen:
                    raise ValueError(
                        f"duplicate event id {event.event_id!r}. Feature rows are indexed "
                        "by it, so one row would be scored against two labels and nothing "
                        "would report it."
                    )
                seen.add(event.event_id)

                examples.append(
                    AttentionExample(
                        domain=event.domain,
                        label=Label(
                            target=VISIT_ENGAGED_TARGET,
                            subject=event.category,
                            window_end=event.occurred_at,
                            outcome=dwell > threshold,
                            horizon_hours=0.0,
                            # The session, not the visit. D94 makes the session the
                            # resampling unit; it does not touch the label's meaning.
                            session_id=session.session_id,
                            label_id=event.event_id,
                        ),
                        row=FeatureRow(
                            subject=event.category,
                            window_end=event.occurred_at,
                            feature_set=feature_set,
                            compat="full",
                            values={name: values[name] for name in names},
                        ),
                    )
                )

            # Only now does this visit become history.
            if dwell is not None:
                history.setdefault(event.category, []).append(dwell)
                domain_dwell.setdefault(event.domain, []).append(dwell)
            domain_visits[event.domain] = domain_visits.get(event.domain, 0) + 1
            domain_days.setdefault(event.domain, set()).add(event.occurred_at.date())
            recent_domains.append(event.domain)
            previous_category = event.category
            previous_domain = event.domain
            previous_dwell = dwell

    return examples

File: extension/src/features/attention.ts
/**
 * `visit_engaged` — will this visit hold you? One label and one `as_2` row per visit.
 *
 * PARITY-CRITICAL: `research/tise_research/features/attention.py` is the mirror, and the
 * two must produce identical labels and identical feature maps within 1e-9.
 *
 * > Given a page you have just opened, will you stay longer than you usually stay on pages
 * > of this category?
 *
 * Adopted in D97 on the author's own browsing, then **replicated across 1,326 other
 * people** in D100, where it beat a constant for 88.6% of them individually.
 *
 * **Dwell arrives here and is never stored.** `TiseEvent.dwellSeconds` is `null` on every
 * row the storage layer will accept — `assertStorable` throws otherwise, and D35 is why.
 * D96 put attention spans in their own store precisely because an event is final when
 * written while a span is opened, extended and closed later. So dwell is joined onto events
 * **in memory**, immediately before features are computed, and the joined value never goes
 * back to disk. The type permits it, storage forbids it, and this module requires it.
 *
 * **Leakage is structural, not checked.** A visit's label and features are emitted *before*
 * that visit joins any state, so a visit can never appear in its own median. The same shape
 * as `block_labels.py`, and the reason no `windowEnd` filter is needed here.
 *
 * **A visit with no dwell is skipped, never imputed.** A missing duration is not a short
 * one, and inventing a value is the defect D51 forbids.
 */
import type { TiseEvent } from "../types";
import { sessionise } from "./sessions";
import { AS2_FEATURE_NAMES, type CompatClass, type FeatureRow, saturate } from "./vector";

/** The target's name. Matches `Prediction.target` in SPEC.md and Python's constant. */
export const VISIT_ENGAGED_TARGET = "visit_engaged";

export const AS2_FEATURE_SET = "as_2";

/** Visits of the same category the median is taken over. */
export const DEFAULT_TRAILING_VISITS = 20;

/** Visits of a category before it can be predicted at all. */
export const DEFAULT_MIN_PRIOR_VISITS = 10;

/**
 * Dwell at which `dwellLevel` reaches 0.5, in seconds. Thirty seconds is an ordinary page
 * read. **Declared, not fitted** (D81), and it does not move per corpus: D100 kept it at 30
 * against a panel whose median dwell was 8s, because a declared constant re-tuned per
 * dataset is a fitted one.
 */
export const DWELL_SCALE_SECONDS = 30;

/** Session position saturation, in visits. Declared. */
export const POSITION_SCALE = 10;

/**
 * Prior visits at which `domainVisits` reaches 0.5. Twenty, matching the trailing window
 * the median is taken over, so there is one notion of "enough history" and not two.
 */
export const DOMAIN_VISIT_SCALE = 20;

/**
 * Visits `domainShare` is taken over. A hundred is a few days of browsing — long enough
 * that a share is not one visit's worth of noise, short enough that a domain abandoned last
 * month has left it.
 */
export const DOMAIN_SHARE_WINDOW = 100;

/** `isDailyDomain` looks back this many days and fires at this many of them. */
export const DAILY_DOMAIN_WINDOW_DAYS = 7;
export const DAILY_DOMAIN_MIN_DAYS = 4;

export interface AttentionLabel {
  readonly target: string;
  readonly subject: string;
  readonly windowEnd: string;
  readonly outcome: boolean;
  readonly horizonHours: number;
  readonly sessionId: string;
  /**
   * Unique per example. Consumers map a label back to its feature row, and keying that on
   * `(subject, windowEnd)` is unique on one person's browsing and **not** in general —
   * 0.02% of the D100 panel's rows are a second visit by one person in the same second.
   */
  readonly labelId: string;
}

export interface AttentionExample {
  readonly label: AttentionLabel;
  readonly row: FeatureRow;
  /**
   * Carried so a per-domain baseline can be fitted, which D24 makes mandatory once four
   * features read the domain. Never displayed and never stored.
   */
  readonly domain: string;
}

export interface AttentionOptions {
  readonly timeoutSeconds: number;
  readonly trailing?: number;
  readonly minPrior?: number;
}

/**
 * Median of a list, mirroring Python's `statistics.median` exactly.
 *
 * For an even count Python averages the two middle values; taking the lower would shift
 * every threshold by half a step and no test on odd-length data would notice.
 */
export function median(values: readonly number[]): number {
  if (values.length === 0) throw new Error("median of no values");
  const ordered = [...values].sort((a, b) => a - b);
  const middle = Math.floor(ordered.length / 2);
  if (ordered.length % 2 === 1) return ordered[middle] as number;
  return ((ordered[middle - 1] as number) + (ordered[middle] as number)) / 2;
}

/**
 * How the person arrived. The one genuinely *conditional* fact each visit carries.
 *
 * `typed` and `auto_bookmark` are deliberate acts; `link` is incidental. The distinction
 * entered the project in D93 and immediately became the second-largest coefficient.
 */
function transitionFlags(transition: string): Record<string, number> {
  const core = transition.toLowerCase();
  return {
    arrivedTyped: core === "typed" || core === "generated" || core === "keyword" ? 1 : 0,
    arrivedBookmark: core.includes("bookmark") ? 1 : 0,
    arrivedLink: core === "link" ? 1 : 0,
  };
}

/**
 * UTC calendar day, as `YYYY-MM-DD`.
 *
 * **UTC, like every other calendar feature in this project** (`context.ts` sets the
 * convention and says what it costs). Python's mirror reads the datetime's own zone, and
 * the events on both sides of the parity fixture are UTC, so the two agree. Using local
 * time here would have shifted `isDailyDomain` and the hour features by the machine's
 * offset and failed nothing except the fixture.
 */
function dayKey(at: Date): string {
  const year = at.getUTCFullYear();
  const month = `${at.getUTCMonth() + 1}`.padStart(2, "0");
  const day = `${at.getUTCDate()}`.padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function daysBefore(at: Date, days: number): string {
  const shifted = new Date(at.getTime());
  shifted.setUTCDate(shifted.getUTCDate() - days);
  return dayKey(shifted);
}

/**
 * One example per visit that has enough history of its own category.
 *
 * `events` must carry `dwellSeconds` joined from the attention store; a visit whose dwell
 * is `null` contributes to history but never receives a label.
 */
export function attentionExamples(
  events: readonly TiseEvent[],
  options: AttentionOptions,
): AttentionExample[] {
  const trailing = options.trailing ?? DEFAULT_TRAILING_VISITS;
  const minPrior = options.minPrior ?? DEFAULT_MIN_PRIOR_VISITS;
  const sessions = sessionise(events, options.timeoutSeconds);

  const history = new Map<string, number[]>();
  const domainDwell = new Map<string, number[]>();
  const domainVisits = new Map<string, number>();
  const domainDays = new Map<string, Set<string>>();
  const recentDomains: string[] = [];

  const examples: AttentionExample[] = [];
  const seen = new Set<string>();
  let previousCategory: string | null = null;
  let previousDomain: string | null = null;
  let previousDwell: number | null = null;

  for (const session of sessions) {
    session.events.forEach((event, position) => {
      const dwell = event.dwellSeconds;
      const prior = history.get(event.category) ?? [];
      const window = prior.slice(-trailing);

      if (dwell !== null && dwell !== undefined && prior.length >= minPrior) {
        const threshold = median(window);
        const at = new Date(event.occurredAt);
        const hourAngle = (2 * Math.PI * at.getUTCHours()) / 24;

        // Domain state, all of it strictly before this visit.
        const domainWindow = (domainDwell.get(event.domain) ?? []).slice(-trailing);
        const firstDay = daysBefore(at, DAILY_DOMAIN_WINDOW_DAYS - 1);
        let daysRecently = 0;
        for (const day of domainDays.get(event.domain) ?? []) {
          if (day >= firstDay) daysRecently += 1;
        }
        if (recentDomains.length === 0) {
          throw new Error("a labelled visit always has prior visits");
        }
        let sameDomainRecently = 0;
        for (const name of recentDomains) {
          if (name === event.domain) sameDomainRecently += 1;
        }

        let windowTotal = 0;
        for (const value of window) windowTotal += value;

        const values: Record<string, number | null> = {
          dwellLevel: saturate(threshold, DWELL_SCALE_SECONDS),
          lastDwellRatio: (window[window.length - 1] as number) / (threshold + 1),
          meanRatio: windowTotal / window.length / (threshold + 1),
          sessionPosition: saturate(position, POSITION_SCALE),
          isSessionStart: position === 0 ? 1 : 0,
          sameAsPrevious: previousCategory === event.category ? 1 : 0,
          hourSin: Math.sin(hourAngle),
          hourCos: Math.cos(hourAngle),
          isWeekend: (at.getUTCDay() + 6) % 7 >= 5 ? 1 : 0,
          ...transitionFlags(event.transition),
          domainVisits: saturate(domainVisits.get(event.domain) ?? 0, DOMAIN_VISIT_SCALE),
          domainShare: sameDomainRecently / recentDomains.length,
          isDailyDomain: daysRecently >= DAILY_DOMAIN_MIN_DAYS ? 1 : 0,
          domainDwellLevel:
            domainWindow.length > 0
              ? saturate(median(domainWindow), DWELL_SCALE_SECONDS)
              : null,
          prevSameDomain: previousDomain === event.domain ? 1 : 0,
          prevDwellRatio: previousDwell === null ? null : previousDwell / (threshold + 1),
        };

        if (seen.has(event.eventId)) {
          throw new Error(
            `duplicate event id ${event.eventId}. Feature rows are indexed by it, so one ` +
              "row would be scored against two labels and nothing would report it.",
          );
        }
        seen.add(event.eventId);

        const selected: Record<string, number | null> = {};
        for (const name of AS2_FEATURE_NAMES) selected[name] = values[name] ?? null;

        examples.push({
          domain: event.domain,
          label: {
            target: VISIT_ENGAGED_TARGET,
            subject: event.category,
            windowEnd: event.occurredAt,
            outcome: dwell > threshold,
            horizonHours: 0,
            sessionId: session.sessionId,
            labelId: event.eventId,
          },
          row: {
            subject: event.category,
            windowEnd: event.occurredAt,
            featureSet: AS2_FEATURE_SET,
            compat: "full" satisfies CompatClass,
            values: selected,
          },
        });
      }

      // Only now does this visit become history.
      if (dwell !== null && dwell !== undefined) {
        const forCategory = history.get(event.category);
        if (forCategory) forCategory.push(dwell);
        else history.set(event.category, [dwell]);

        const forDomain = domainDwell.get(event.domain);
        if (forDomain) forDomain.push(dwell);
        else domainDwell.set(event.domain, [dwell]);
      }
      domainVisits.set(event.domain, (domainVisits.get(event.domain) ?? 0) + 1);
      const days = domainDays.get(event.domain);
      const key = dayKey(new Date(event.occurredAt));
      if (days) days.add(key);
      else domainDays.set(event.domain, new Set([key]));

      recentDomains.push(event.domain);
      if (recentDomains.length > DOMAIN_SHARE_WINDOW) recentDomains.shift();

      previousCategory = event.category;
      previousDomain = event.domain;
      previousDwell = dwell ?? null;
    });
  }

  return examples;
}
